seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,227 @@
+    if root_url is None:
+        root_url = TERRA_API_ROOT_URL
+    url = urljoin(root_url, methcall)
+    return url, headers
+
+
+class AnvilSession(AuthorizedSession):
+
+    def __init__(self, credentials=None, service_account_info=None, scopes=None):
+        """
+        Create an AnVIL session for a user account if credentials are provided, otherwise create one for the service account
+
+        :param credentials: User credentials
+        :param service_account_info: service account secrects
+        :param scopes: scopes of the access privilege of the session
+        """
+        if credentials is None:
+            credentials = service_account.Credentials.from_service_account_info(service_account_info, scopes = scopes)
+        super(AnvilSession, self).__init__(credentials)
+
+    def get(self, methcall, headers=None, root_url=None, **kwargs):
+        """
+        Call Terra API with HTTP GET method with an authentication header.
+
+        Args:
+            methcall: A string of API path (start right after the domain name without leading slash (/)
+            headers (dict): Include additional headers as key-value pairs
+            root_url: the url up to the domain name ending with a slash (/)
+            kwargs: other parameters for the HTTP call, e.g. queries.
+        Return:
+            HTTP response
+        """
+        url, headers = _get_call_args(methcall, headers, root_url)
+        r = super(AnvilSession, self).get(url, headers = headers, **kwargs)
+        logger.info('GET {} {} {}'.format(url, r.status_code, len(r.text)))
+        return r
+
+    def post(self, methcall, headers=None, root_url=None, **kwargs):
+        """See the __get() method."""
+        url, headers = _get_call_args(methcall, headers, root_url)
+        r = super(AnvilSession, self).post(url, headers = headers, **kwargs)
+        logger.info('POST {} {} {}'.format(url, r.status_code, len(r.text)))
+        return r
+
+    def put(self, methcall, headers=None, root_url=None, **kwargs):
+        """See the __get() method."""
+        url, headers = _get_call_args(methcall, headers, root_url)
+        r = super(AnvilSession, self).put(url, headers = headers, **kwargs)
+        logger.info('PUT {} {} {}'.format(url, r.status_code, len(r.text)))
+        return r
+
+    def delete(self, methcall, headers=None, root_url=None):
+        """See the __get() method."""
+        url, headers = _get_call_args(methcall, headers, root_url)
+        r = super(AnvilSession, self).delete(url, headers = headers)
+        logger.info('DELETE {} {} {}'.format(url, r.status_code, len(r.text)))
+        return r
+
+
+def get_anvil_billing_projects(user):
+    """
+    Get activation information for the logged-in user.
+
+    Args:
+        user (User model): who's credentials will be used to access AnVIL
+
+    :returns a list of billing project dictionary
+    """
+    session = anvil_session_store.get_session(user)
```

Comment:
There are checks for the existence of a valid session before calling this function. But maybe checking here is better and well protected.

Reply:
yes, you can't assume a publicly available method is called correctly everywhere

Reply:
Since it is required to check that the user does have a session before calling this function, so we don't check whether the `session` is `None` or not. If `session` is `None` because of something wrong, it will cause an exception in the next statement. Is it the normal way for Python to handle errors?

Reply:
There is an enormous difference between a user who never logged in with google and isn't supposed to have access, and a user where the cache is somehow corrupted or other errors have occured and they should have access but don't have a session. You want to gracefully handle the error if the user isn't supposed to have access to this method, and you want to raise the exception if they do. As I mentioned elsewhere, you need a helper method to check if a user is authenticated through google or not